### ISSUE TITLE
Add proper backend interface to handle user code from frontend

### ISSUE THREAD

**Issue by AndriiSerbeniuk** [ITA-Dnipro/CompilerWebApp#4]
Submit is done by POST HTTP request with input json as its body. Route for the request is ```/submit```.
Response will contain output JSON as its body.

Current input JSON is:
```
{
    code: String,
    options: String 
}
```

Current output JSON is:
```
{
    stdin: String,
    stdout: String,
    result: Blob
}
```
Should failure-case JSON be the same but just with empty result? I don't quite get it yet.

Compiler crate's input and output structs should match the JSONs, I suppose.

***Suggestions for changes are welcome.***

**Comment by virtustron** [comment]
> 
> Should failure-case JSON be the same but just with empty result? I don't quite get it yet.

Upon initial consideration there are possibly such cases:
 - failure with JSON serializing on frontend's side (empty "source code" field, for instance) -> no JSON sending to backend + message to user;
 - failure with JSON deserializing on backend side (corrupted file, invalid fields, etc) -> no compiler calling + message to frontend.

But if you have other issues - it should be different result.


> Compiler crate's input and output structs should match the JSONs, I suppose.

There are various options, to my mind. Maybe, it will be useful to add some extra data to JSON - only for frontend-backend "communication" purposes, not for compiler.
In [data folder](https://github.com/ITA-Dnipro/CompilerWebApp/tree/feature/compiler/compiler/src/data) you can find my proposals on data structures - suggestions will be very welcome

**Comment by AndriiSerbeniuk** [comment]
> Maybe, it will be useful to add some extra data to JSON - only for frontend-backend "communication" purposes, not for compiler.

If need be this can be done with request headers, I don't think we'll need body for it. And requests body is only supposed to carry working data anyway, as far as I know.


> Compiler crate's input and output structs should match the JSONs, I suppose.

This would be cool for easy parsing of input request into struct directly. Ideally I'd like to implement Serialize and Deserialize traits for the structs in the compiler module. On the backend side of course, I don't think there's need to bother compiler crate with it.
This would skip transition steps between raw request body and the structure that the compiler needs, rocket would parse json into it directly, without excessive data copying between transitioning structures. 

But this is not that important, just if it so happened to be possible it'd be cool.

**Comment by AndriiSerbeniuk** [comment]
> Current input JSON is:
> 
> ```
> {
>     code: String,
>     options: String 
> }
> ```

This structure is just my current working placeholder, ideally I'd like to use compilation crate's input and output data structures.
For now I'll add language field to it and go with until we are sure about how they should look like.

**Comment by AndriiSerbeniuk** [comment]
> > Compiler crate's input and output structs should match the JSONs, I suppose.
> 
> This would be cool for easy parsing of input request into struct directly. Ideally I'd like to implement Serialize and Deserialize traits for the structs in the compiler module.

After some discussions I deemed this improbable. I just now realized that compiler needs a file with source code as its input and that I need to extract it from the JSON. Transition between source code as String and as a File will have to be done anyway. I'm leavng this idea, at least for now. 
Not binding to compiler module's structures so strongly gives backend more freedom of input data preprocessing, so I guess it's a good thing at the end of the day.
